design-proposals/klip-20_remove_terminate.md

```diff
@@ -0,0 +1,121 @@
+# KLIP 20 - Remove 'TERMINATE' statements
+
+**Author**: @big-andy-coates | 
+**Release Target**: TBD | 
+**Status**: In Discussion | 
+**Discussion**: TBD
+
+**tl;dr:** 
+Requiring users to `TERMINATE` a persistent query before dropping a source causes friction in the 
+user experience. The id of the query(s) that need terminating are not determinist or accessible 
+through SQL. We propose dropping the source should automatically drop persistent query(s) that write
+to it.
+
+*NB*: is most likely dependent on [KLIP-17](klip-17-sql-union.md).
+
+## Motivation and background
+
+See https://github.com/confluentinc/ksql/issues/3967 and the issues it links to for more background.
+
+In short, dropping streams and tables created using `CSAS` or `CTAS` is currently painful, and 
+practically impossible from a script. 
+
+If done manually, the user must run `DESCRIBE <source>` to get the list of queries writing to the 
+source and then stop each of these in turn, via a `TERMINATE` statement, before finally being able 
+to drop the stream or table. 
+
+From a script, it is not possible to determine the id of the queries that need to be stopped, and 
+those query ids will change from one run to another in most cases. Hence it is generally not 
+possible to drop a table or stream in a script.
+
+The problem is compounded by the fact that _mutiple_ persistent queries can be writing into the
+source or table.  However, [KLIP-17](klip-17-sql-union.md) proposes to correct this, so that there
+will be, at-most, a single persistent query writing to any table or stream.
+
+There also seems to be no clear benefit of being able to terminate a query without dropping the 
+associated stream or table. Especially, given there is no means to restart it.
+
+Consider that `CSAS` and `CTAS` statements are equivalent to `MATERIALIZED VIEW`s used in RDBSs, 
+(as proposed in [KLIP-19](klip-19-materialize-views.md)). In a traditional RDBS the underlying 
+process by which the materialized view is kept up to data is not exposed to the user.  There is
+no 'persistent query' to either list or terminate. Hence we propose we should remove the 
+problematic `TERMINATE` from our syntax and instead allow users to drop the view, even if there 
+are queries writing to it. If such queries exist they will simply be stop as part of the `DROP`.   
+
+## What is in scope
+
+* Removing `TERMINATE` from our SQL syntax.
```

Comment:
What's the benefit of removing `TERMINATE`. Can't we just drop the requirement that the query be terminated to drop the stream/table?

Reply:
We cannot drop a stream/table if a query is writing into it. Otherwise we will have queries that write into zombie stream/tables!

Reply:
@hjafarpour - I think @rodesai means to automatically discover and terminate the queries writing into it

Reply:
Why not drop `TERMINATE` ? What use is it?  Trad RDBS doesn't expose the persistent query updating materialized views to users.  It's an implementation detail that shouldn't be exposed IMHO.

Reply:
I agree that dropping a MV, should imply that the corresponding query is stopped. However, I think it would still be useful to allow users to terminate a persistent query directly. I also commented on KLIP-17, as I think, keeping `INSERT INTO` as-is is useful.

Reply:
> This KLIP seems to relate to the "ownership model" that ksqlDB is lacking

I don't see how it is...

> If we assume that the output topic is not owned by ksqlDB, but a query write into it, it make total sense to allow users to just terminate the query.

To be able to have a persistent query that writes to the topic there must be a source registered in KSQL that is backed by the topic.  Currently this would either be by a `CREATE STREAM` or `CREATE STREAM AS SELECT` statement.  Neither of these currently embody any sense of ownership. 

Dropping the source created by the `CS` or `CSAS` doesn't currently delete the topic, and that wouldn't change with this KLIP.

What this KLIP proposes is that if the user wants to stop the query they would just drop the source.  The query is an implementation detail.